### ISSUE TITLE
Support dynamic retry delays

### ISSUE THREAD

**Issue by whiskeysierra** [jhalterman/failsafe#110]
Server's that use rate limiting often use [`429 Too Many Requests`](https://httpstatuses.com/429) that includes a `Retry-After` header. It would be nice if Failsafe would support to dynamically set a delay for one particular exception.

**Comment by Tembrel** [comment]
I was going to argue that you could roll your own version of this, but it turned out to be a lot harder than I thought. Implementing it within Failsafe is more feasible, so I created PR #111 as a proposal that I believe would address your request.

**Comment by Tembrel** [comment]
Using the `RetryPolicy.withDelayFunction` from #111, your example could be written something like this:
```java
CompletableFuture<Response> makeRequest(Request request)
        throws TooManyRequestsException {
    return ...;
}

CompletableFuture<Response> makeFailsafeRequest(Request request) {
    RetryPolicy retryPolicy = new RetryPolicy()
	.retryOn(TooManyRequestsException.class)
	.withDelayFunction((response, exception) -> {
	    if (exception instanceof TooManyRequestsException) {
		TooManyRequestsException tmre =
                    (TooManyRequestsException) exception;
		long retryAfterSeconds = tmre.getRetryAfterSeconds();
		if (retryAfterSeconds >= 0) {
		    return new Duration(
                        retryAfterSeconds, TimeUnit.SECONDS);
		}
	    }
	    return null;
	});

    return Failsafe.with(retryPolicy)
	.with(schedExecSvc)
	.future(() -> makeRequest(request));
}
```

**Comment by whiskeysierra** [comment]
Even if the exception is not thrown directly but if the future ends
exceptionally?

**Comment by Tembrel** [comment]
Right, in fact, I probably shouldn't have put `throws TooManyRequestsException` in the signature of `makeRequest` and just said that the resulting `CompletableFuture` could complete exceptionally with that exception.

You could write a little test to confirm that that would work (without using the proposed new method).

**Comment by jhalterman** [comment]
Good use case. Thanks for filing this.

**Comment by jhalterman** [comment]
This is released in 1.1.0.
